contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,205 @@
+package tracing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"go.temporal.io/sdk/interceptor"
+	"go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/workflow"
+)
+
+var (
+	_ interceptor.Tracer     = tracerImpl{}
+	_ interceptor.TracerSpan = &tracerSpan{}
+)
+
+// NewInterceptor creates an interceptor for setting on client options
+// that implements Datadog tracing for workflows.
+func NewInterceptor() interceptor.Tracer {
+	return tracerImpl{}
+}
+
+type contextKey string
+
+const (
+	activeSpanContextKey contextKey = "dd_trace_span"
+	headerKey                       = string(activeSpanContextKey)
+)
+
+type tracerImpl struct {
+	interceptor.BaseTracer
+}
+
+func (tracerImpl) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: activeSpanContextKey,
+		HeaderKey:      headerKey,
+	}
+}
+
+func (tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func genSpanID(runId string) uint64 {
```

Comment:
Just in case there end up being multiple spans which require deterministic IDs, and also to avoid user code for custom spans accidentally colliding, I think we should include a bit more than just the run ID as input to the hashing function. Perhaps something like `<workflow_namespace>:<workflow_id>:<workflow_run_id>:<trace_operation_name>` would be more future proof?

Reply:
This is also no longer a concern as a result of #931.

Reply:
Updated to use it!